### PR TITLE
Contact Sorting

### PR DESCRIPTION
**Pull request by tylerhm**
Contacts are now sorted with a modified bubble sort. If there is a better way to sort html elements let me know, as bubble sort was the only one I could apply to this workflow.

### CODE REVIEW

File: js/contact.js [hunk 1]
```diff
@@ -281,4 +284,31 @@ function appendContactLink(contact) {
 	contactLanding.appendChild(contactLink);
 
 	return contactLink;
+}
+
+/**
+ * Sorts all contacts by first name
+ */
+function sortContacts() {
+	let contacts = document.getElementById("contactLanding");
```
Comment (suneettipirneni) [inline js/contact.js]: Luckily, you don't need to know any sorting algorithms to sort things in js. All you need is a comparator, or in other words, something to tell js what is *greater* and what is *less* and it'll do the rest for you:

```javascript
// Sorting by object's name
myCollection.sort((obj1, obj2) => (obj1.name > obj2.name) ? 1 : -1);
```

When it comes to content of sorting, you should sort the contacts map, and then repopulate the DOM based on those values. This saves a lot of DOM class lookups and makes things more concise overall.

**Review by suneettipirneni — CHANGES_REQUESTED** [review]

File: js/contact.js (hunk 1 above)
Comment (tylerhm) [reply js/contact.js]: This is awesome! I had no idea that maps were even sortable. I had always thought they were meant to be an unordered datastructure. Thank you!

Comment (suneettipirneni) [reply js/contact.js]: The maps themselves aren’t sortable but you can sort their contents into an array

Comment (tylerhm) [reply js/contact.js]: Got it :+1:

Comment (suneettipirneni) [reply js/contact.js]: This seems to work for me:

```javascript
const alphaComparator = (contact1, contact2) => (contact1.FirstName > contact2.FirstName) ? 1 : -1;

 // It's a map so it nests its array, hence the [0] subscript.
const sorted = new Array([...contacts.values()].sort(alphaComparator))[0];
```

Comment (tylerhm) [reply js/contact.js]: I have something very similar to this in essence. We are now using the sort function as our default for displaying contacts, as we want them to always be in sorted order. The displaying must also happen AFTER the sorting, which is why I did this.

Comment (tylerhm) [reply js/contact.js]: Unfortunately, this style seems to break the auto selecting feature I built in a while ago, fixing now

Comment (suneettipirneni) [reply js/contact.js]: Maybe decouple sorting from the populating method and pass the sorted array into instead

**Comment by tylerhm** [comment]
Ok the sort is done more 'javascripty' now, and contact selection is maintained.

This new flow of sorting and then displaying will allow seamless search integration, as I can simply pass the values returned from the search directly into the sort and display function :)

**Review by suneettipirneni — CHANGES_REQUESTED** [review]
From my humble perspective PR needs a good amount of refactoring. Which is fine, as I said before this should be a learning process. However, instead of offering a specific line to change I want to explain my reasoning as I've never properly explained my perspective on structuring code:

### Too much behavior defined in a single function is not good

The `populateContacts()` method was originally built to fetch and display contacts without any added functionality. This worked well because it did exactly what it was expected to do no more and no less. I wanted to eventually remove the GET request from it so it would only exclusively edit the DOM. So in essence `populateContacts()` would only do one thing:

- Repopulate the contact list DOM.

The current suggested changes are a huge departure from this. Let's look at how many things the suggested function does under the umbrella of a single function call:

- Does a GET request
- Modifies backing contact map
- Alphabetical Sort data from map
- Modifies Dom with specific sort order
- Runs a select DOM operation on the element.

Some of these bring me to my next point

### Dependency Inversion Principle is a good guideline
7-dependency_inversion_principle


If we bake a bunch of hardcoded functionality into a single function, then if we ever want make more granular changes, it becomes extremely difficult. For example:

"Ok the contact list sorts alphabetically by first name, but what if there's an option to sort by last name? Or maybe both?"

Well `sortAndPopulate()` sorts first-name alphabetically regardless of the input. We could add a parameter to `sortAndPopulate` for a comparator to remedy this issue. However there's another issue:

"Ok I want the contacts to be sorted but I don't want to do a GET request every time it's sorted, after all, we already have the data available, it would be faster". 

Well I guess we could solve this by adding another parameter to `sortAndPopulate` for a boolean that tells us whether to do a request or not. We'll also have to cram more logic into the function to get it to work.

We've now increased the size of our method body, and have doubled our parameters bringing the total to 4 parameters. 

### Avoiding bloated-hardcoded behavior by decoupling logic does wonders

Luckily avoiding the situation above is trivial. Let me propose a different function structure. We'll start by changing the parameters of the `populateContacts` function:
```javascript
    function populateContacts(contacts)
```
Where `contacts` is an array of contacts. This method will take the array *in the order that it was given* and populate contact links for it, nothing more and nothing less.

Believe it or not we already fixed an issues mentioned above. This is because we can pass *any* kind of a sorted array into the method and it will display in that order. 

What about avoiding a GET request every time we call the method? Well we're no longer fetching contacts from within the method we're passing an array in, therefore we know for a fact any GET requests are being done outside of the method.

So what exactly would this look like in code?
```javascript
    window.onload = function() {
	    // Basic HTML setup code...

		// The page has loaded, we need to fetch the contacts.
		const contactsArr = getContacts(id);

		// We want to have a firstname-sort by default.
		const sorted = contactsArr.sort(firstnameComparator);

		// Now that we've sorted this array we need to render it
		populateContacts(sorted);

		// We select the first contact by default so do that
		select(sorted[0]);
    }
```
We can take this a step further and implement user-selectable sorts
```javascript
    $("sort-selection").onChoose(function() {
	    e.preventDefault();
	    // It looks like the user would like to sort another way!
	    
		// (For the sake of brevity lets assume they chose to sort 
		// by last name).	
		
		// No problem, just sort the current array we have now.
		const sorted = contacts.sort(lastnameComparator);
		populateContacts(sorted);
    }
```
Because our code was decoupled we were able to build a more robust and flexible codebase that didn't need to be refactored every time we needed a new feature.

**Comment by tylerhm** [comment]
This is hugely helpful! I've always been a big stickler about encapsulation as well; however, I've found it hard to implement properly without an intimate knowledge of the event driven processes that are used so frequently in JS. I will make these changes, thank you so much for the clear feedback :)

I seriously appreciate the effort that went into this explanation.

**Comment by tylerhm** [comment]
^^ Made some changes to follow the style you described more closely. Feels much more coherent and easy to follow than before. Feedback would be much appreciated!

**Review by suneettipirneni — COMMENTED** [review]
This looks very well done! I think I'll make some minor changes later today but as it stands now, your implementation is extremely robust, nice work!

**Review by suneettipirneni — APPROVED** [review]

**Comment by tylerhm** [comment]
Thank you for the feedback!! I like the comparator change, it's much more concise. I have some free time, so will implement searching in a few hours or so :)

**Pull request merged**